java/org/apache/catalina/session/StandardSession.java

```diff
@@ -1644,6 +1727,27 @@ protected void doWriteObject(ObjectOutputStream stream) throws IOException {
             manager.getContext().getLogger().debug
                 ("writeObject() storing session " + id);
 
+        // Write authentication information if requested, session is authenticated AND
+        // if principal is serializable
+        Principal sessionPrincipal = getPrincipal();
+        if (persistAuthentication && isAuthenticated(sessionPrincipal)
+                && isPrincipalSerializable(sessionPrincipal)) {
+            // Boolean tag: authentication information has been serialized
+            stream.writeObject(Boolean.TRUE);
```

Comment:
Again, just simplify: always write the values and forget the Boolean.

This will be easier if you remove the ability for each session to decide if it wants to persist authentication information. It really should be a property of the Manager and not the session.